gusto/form_manipulation_labelling.py

```diff
@@ -0,0 +1,328 @@
+                isinstance(replacer, ufl.core.expr.Expr)
+                and not isinstance(replacer, Function)
+            )
+
+            # size returns either the length of a tuple or the length
+            # of a function's function space
+            size = lambda q: len(q) if type(q) is tuple else len(q.function_space())
+
+            # if either replacer or labelled are Functions defined on
+            # a mixed function space then split them to get a tuple of
+            # functions
+            if type(replacer) is Function and size(replacer) > 1:
+                new = replacer.split()
+            else:
+                new = replacer
+            if type(labelled) is Function and size(labelled) > 1:
+                old = labelled.split()
+            else:
+                old = labelled
+
+            # constract the replace_dict = {old: new}, comparing the
+            # sizes of the part to be replaced and its replacement to
+            # figure out if we need to extract just one component of
+            # either of them
+            if size(old) == 1:
+                if not repl_expr and size(new) > 1:
+                    new = new[old.function_space().index]
+                replace_dict[old] = new
+            else:
+                if repl_expr or size(new) == 1:
+                    replace_dict[old[new.function_space().index]] = new
+                else:
+                    for k, v in zip(old, new):
+                        replace_dict[k] = v
+
+        # construct new form and term
+        new_form = ufl.replace(t.form, replace_dict)
+        new_term = Term(new_form, t.labels)
+
+        # if we have replaced a labelled part of the form with another
+        # labelled part then we need to remove the original label
+        if remove_labels:
+            for label in labels:
+                new_term = label.remove(new_term)
+
+        return new_term
+
+    return rep
+
+
+class Term(object):
+    """
+    Term class, containing a form and its labels.
+
+    :arg form: the form for this term
+    :arg label_dict: dictionary of key-value pairs corresponding to current form labels.
+    """
+    __slots__ = ["form", "labels"]
+
+    def __init__(self, form, label_dict=None):
+        self.form = form
+        self.labels = label_dict or {}
+
+    def get(self, label, default=None):
+        return self.labels.get(label.label)
+
+    def has_label(self, *labels):
+        if len(labels) == 1:
+            return labels[0].label in self.labels
+        else:
+            return tuple(self.has_label(l) for l in labels)
+
+    def __add__(self, other):
+        if other is None:
+            return self
+        elif isinstance(other, Term):
+            return LabelledForm(self, other)
+        elif isinstance(other, LabelledForm):
+            return LabelledForm(self, *other.terms)
+        else:
+            return NotImplemented
+
+    __radd__ = __add__
+
+    def __rmul__(self, other):
```

Comment:
Wouldn't it be better if here we define the `__mul__` method and just set `__mul__ = __rmul__`? I guess in the end, Python will first try to call `__mul__` and notice it doesn't have one, then the right argument's `__rmul__` will be called.

Reply:
I'm happy to do that if it's the right way - I don't have much experience with these magic methods.

Reply:
Yes. That's the right thing to do.